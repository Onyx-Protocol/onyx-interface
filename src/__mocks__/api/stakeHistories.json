{
  "data": [
    {
      "id": 291,
      "address": "0x0000000000000000000000000000000000000000",
      "tx_hash": "0xf97a7673403a643ad23edb9ffa4c60bf4192998aaf436782240b84be94d3aba1",
      "block_hash": "0x7d6991ef8f031e9cdc4ced2f924dee008e505c741f169101124aee16dad795f1",
      "type": 0,
      "amount": "11075000000000000000",
      "reward": "0",
      "price": "0.033695332045177356",
      "created_at": "2022-12-14T19:04:11.000Z",
      "updated_at": "2022-12-14T19:04:11.000Z"
    }
  ],
  "metadata": {
    "page": 1,
    "limit": 5,
    "totalItem": 1,
    "totalPage": 1
  }
}
