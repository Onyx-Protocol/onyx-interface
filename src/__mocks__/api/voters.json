{
  "data": [
    {
      "id": "40b3dcb3-f337-4713-ae4c-cae927351550",
      "address": "0xe864c9439eb042e969c59dd13d38cfdb28e16d73",
      "has_voted": 1,
      "support": 1,
      "votes": "232456299820100000000000000",
      "proposal_id": 6,
      "block_number": 14636926,
      "block_timestamp": 1650660975,
      "created_at": "2022-07-11T01:00:47.233Z",
      "updated_at": "2022-07-11T01:01:00.000Z"
    }
  ],
  "metadata": {
    "sumVotes": "232456299820100000000000000",
    "page": 1,
    "limit": 4,
    "totalItem": 1,
    "totalPage": 1
  }
}