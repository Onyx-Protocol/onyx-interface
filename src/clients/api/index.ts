export { default as queryClient } from './queryClient';

// Mutations
export { default as enterMarkets } from './mutations/enterMarkets';
export * from './mutations/enterMarkets';
export { default as useEnterMarkets } from './mutations/enterMarkets/useEnterMarkets';

export { default as exitMarket } from './mutations/exitMarket';
export * from './mutations/exitMarket';
export { default as useExitMarket } from './mutations/exitMarket/useExitMarket';

export { default as approveToken } from './mutations/approveToken';
export * from './mutations/approveToken';
export { default as useApproveToken } from './mutations/approveToken/useApproveToken';

export { default as approveNftForAll } from './mutations/approveNftForAll';
export * from './mutations/approveNftForAll';
export { default as useApproveNftForAll } from './mutations/approveNftForAll/useApproveNftForAll';

export { default as supplyNonEth } from './mutations/supplyNonEth';
export * from './mutations/supplyNonEth';
export { default as useSupplyNonEth } from './mutations/supplyNonEth/useSupplyNonEth';
export * from './mutations/supplyNonEth/useSupplyNonEth';

export { default as supplyEth } from './mutations/supplyEth';
export * from './mutations/supplyEth';
export { default as useSupplyEth } from './mutations/supplyEth/useSupplyEth';
export * from './mutations/supplyEth/useSupplyEth';

export { default as supplyNFT } from './mutations/supplyNFT';
export * from './mutations/supplyNFT';
export { default as useSupplyNFT } from './mutations/supplyNFT/useSupplyNFT';
export * from './mutations/supplyNFT/useSupplyNFT';

export { default as redeem } from './mutations/redeem';
export * from './mutations/redeem';
export { default as useRedeem } from './mutations/redeem/useRedeem';

export { default as redeemNFT } from './mutations/redeemNFT';
export * from './mutations/redeemNFT';
export { default as useRedeemNFT } from './mutations/redeemNFT/useRedeemNFT';

export { default as repayNonEthOToken } from './mutations/repayNonEthOToken';
export * from './mutations/repayNonEthOToken';
export { default as useRepayNonEthOToken } from './mutations/repayNonEthOToken/useRepayNonEthOToken';

export { default as repayEth } from './mutations/repayEth';
export * from './mutations/repayEth';
export { default as useRepayEth } from './mutations/repayEth/useRepayEth';

export { default as redeemUnderlying } from './mutations/redeemUnderlying';
export * from './mutations/redeemUnderlying';
export { default as useRedeemUnderlying } from './mutations/redeemUnderlying/useRedeemUnderlying';

export { default as claimXcnReward } from './mutations/claimXcnReward';
export * from './mutations/claimXcnReward';
export { default as useClaimXcnReward } from './mutations/claimXcnReward/useClaimXcnReward';

export { default as borrowOToken } from './mutations/borrowOToken';
export * from './mutations/borrowOToken';
export { default as useBorrowOToken } from './mutations/borrowOToken/useBorrowOToken';

export { default as useRepayOToken } from './mutations/useRepayOToken';

export { default as createProposal } from './mutations/createProposal';
export * from './mutations/createProposal';
export { default as useCreateProposal } from './mutations/createProposal/useCreateProposal';

export { default as cancelProposal } from './mutations/cancelProposal';
export * from './mutations/cancelProposal';
export { default as useCancelProposal } from './mutations/cancelProposal/useCancelProposal';

export { default as executeProposal } from './mutations/executeProposal';
export * from './mutations/executeProposal';
export { default as useExecuteProposal } from './mutations/executeProposal/useExecuteProposal';

export { default as queueProposal } from './mutations/queueProposal';
export * from './mutations/queueProposal';
export { default as useQueueProposal } from './mutations/queueProposal/useQueueProposal';

export { default as castVote } from './mutations/vote/castVote';
export * from './mutations/vote/castVote';
export { default as useCastVote } from './mutations/vote/useCastVote';

export { default as registerProxy } from './mutations/registerProxy';
export * from './mutations/registerProxy';
export { default as useRegisterProxy } from './mutations/registerProxy/useRegisterProxy';

export { default as depositPunk } from './mutations/depositPunk';
export * from './mutations/depositPunk';
export { default as useDepositPunk } from './mutations/depositPunk/useDepositPunk';

export { default as mintWPunks } from './mutations/mintWPunks';
export * from './mutations/mintWPunks';
export { default as useMintWPunks } from './mutations/mintWPunks/useMintWPunks';

export { default as burnWPunks } from './mutations/burnWPunks';
export * from './mutations/burnWPunks';
export { default as useBurnWPunks } from './mutations/burnWPunks/useBurnWPunks';

export { default as swapTokens } from './mutations/swapTokens';
export * from './mutations/swapTokens';
export { default as useSwapTokens } from './mutations/swapTokens/useSwapTokens';

// Queries
export { default as getAssetsInAccount } from './queries/getAssetsInAccount';
export * from './queries/getAssetsInAccount';
export { default as useGetAssetsInAccount } from './queries/getAssetsInAccount/useGetAssetsInAccount';

export { default as getHypotheticalAccountLiquidity } from './queries/getHypotheticalAccountLiquidity';
export * from './queries/getHypotheticalAccountLiquidity';

export { default as getMarkets } from './queries/getMarkets';
export * from './queries/getMarkets';
export { default as useGetMarkets } from './queries/getMarkets/useGetMarkets';

export { default as getOTokenBalancesAll } from './queries/getOTokenBalancesAll';
export * from './queries/getOTokenBalancesAll';
export { default as useGetOTokenBalancesAll } from './queries/getOTokenBalancesAll/useGetOTokenBalancesAll';

export { default as getOTokenBalanceOf } from './queries/getOTokenBalanceOf';
export * from './queries/getOTokenBalanceOf';
export { default as useGetOTokenBalanceOf } from './queries/getOTokenBalanceOf/useGetOTokenBalanceOf';

export { default as getXcnReward } from './queries/getXcnReward';
export * from './queries/getXcnReward';
export { default as useGetXcnReward } from './queries/getXcnReward/useGetXcnReward';

export { default as getAllowance } from './queries/getAllowance';
export * from './queries/getAllowance';
export { default as useGetAllowance } from './queries/getAllowance/useGetAllowance';

export { default as getIsApprovedForAll } from './queries/getIsApprovedForAll';
export * from './queries/getIsApprovedForAll';
export { default as useGetIsApprovedForAll } from './queries/getIsApprovedForAll/useGetIsApprovedForAll';

export { default as getBalanceOf } from './queries/getBalanceOf';
export * from './queries/getBalanceOf';
export { default as useGetBalanceOf } from './queries/getBalanceOf/useGetBalanceOf';

export { default as getTokenBalances } from './queries/getTokenBalances';
export * from './queries/getTokenBalances';
export { default as useGetTokenBalances } from './queries/getTokenBalances/useGetTokenBalances';

export { default as useGetUserMarketInfo } from './queries/useGetUserMarketInfo';

export { default as useGetTreasuryTotals } from './queries/useGetTreasuryTotals';

export { default as getMarketHistory } from './queries/getMarketHistory';
export * from './queries/getMarketHistory';
export { default as useGetMarketHistory } from './queries/getMarketHistory/useGetMarketHistory';

export { default as getOTokenCash } from './queries/getOTokenCash';
export * from './queries/getOTokenCash';
export { default as useGetOTokenCash } from './queries/getOTokenCash/useGetOTokenCash';

export { default as getOTokenInterestRateModel } from './queries/getOTokenInterestRateModel';
export * from './queries/getOTokenInterestRateModel';
export { default as useGetOTokenInterestRateModel } from './queries/getOTokenInterestRateModel/useGetOTokenInterestRateModel';

export { default as getOTokenApySimulations } from './queries/getOTokenApySimulations';
export * from './queries/getOTokenApySimulations';
export { default as useGetOTokenApySimulations } from './queries/getOTokenApySimulations/useGetOTokenApySimulations';

export { default as getCurrentVotes } from './queries/getCurrentVotes';
export * from './queries/getCurrentVotes';
export { default as useGetCurrentVotes } from './queries/getCurrentVotes/useGetCurrentVotes';

export { default as getStakingInfos } from './queries/getStakingInfos';
export * from './queries/getStakingInfos';
export { default as useGetStakingInfos } from './queries/getStakingInfos/useGetStakingInfos';

export { default as getStakingApy } from './queries/getStakingApy';
export * from './queries/getStakingApy';
export { default as useGetStakingApy } from './queries/getStakingApy/useGetStakingApy';

export { default as getTransactions } from './queries/getTransactions';
export * from './queries/getTransactions';
export { default as useGetTransactions } from './queries/getTransactions/useGetTransactions';

export { default as getDailyXcn } from './queries/getDailyXcn';
export * from './queries/getDailyXcn';
export { default as useGetDailyXcn } from './queries/getDailyXcn/useGetDailyXcn';

export { default as getProposals } from './queries/getProposals';
export * from './queries/getProposals';
export { default as useGetProposals } from './queries/getProposals/useGetProposals';

export * from './queries/getStakeHistories';
export { default as useGetStakeHistories } from './queries/getStakeHistories/useGetStakeHistories';

export { default as getProposal } from './queries/getProposals/getProposal';
export * from './queries/getProposals/getProposal';
export { default as useGetProposal } from './queries/getProposals/useGetProposal';

export { default as getVoteReceipt } from './queries/getVoteReceipt';
export * from './queries/getVoteReceipt';
export { default as useGetVoteReceipt } from './queries/getVoteReceipt/useGetVoteReceipt';

export { default as getVoters } from './queries/getVoters';
export * from './queries/getVoters';
export { default as useGetVoters } from './queries/getVoters/useGetVoters';

export { default as getVoterDetails } from './queries/getVoterDetails';
export * from './queries/getVoterDetails';
export { default as useGetVoterDetails } from './queries/getVoterDetails/useGetVoterDetails';

export { default as getVoterHistory } from './queries/getVoterHistory';
export * from './queries/getVoterHistory';
export { default as useGetVoterHistory } from './queries/getVoterHistory/useGetVoterHistory';

export { default as getVoterAccounts } from './queries/getVoterAccounts';
export * from './queries/getVoterAccounts';
export { default as useGetVoterAccounts } from './queries/getVoterAccounts/useGetVoterAccounts';

export { default as getProposalThreshold } from './queries/getProposalThreshold';
export * from './queries/getProposalThreshold';
export { default as useGetProposalThreshold } from './queries/getProposalThreshold/useGetProposalThreshold';

export { default as getProposalState } from './queries/getProposalState';
export * from './queries/getProposalState';
export { default as useGetProposalState } from './queries/getProposalState/useGetProposalState';

export { default as getLatestProposalIdByProposer } from './queries/getLatestProposalIdByProposer';
export * from './queries/getLatestProposalIdByProposer';
export { default as useGetLatestProposalIdByProposer } from './queries/getLatestProposalIdByProposer/useGetLatestProposalIdByProposer';

export { default as getBlockNumber } from './queries/getBlockNumber';
export * from './queries/getBlockNumber';
export { default as useGetBlockNumber } from './queries/getBlockNumber/useGetBlockNumber';

export { default as getProposalEta } from './queries/getProposalEta';
export * from './queries/getProposalEta';
export { default as useGetProposalEta } from './queries/getProposalEta/useGetProposalEta';

export { default as getProxies } from './queries/getProxies';
export * from './queries/getProxies';
export { default as useGetProxies } from './queries/getProxies/useGetProxies';

export { default as getOwnedPunkIds } from './queries/getOwnedPunkIds';
export * from './queries/getOwnedPunkIds';
export { default as useGetOwnedPunkIds } from './queries/getOwnedPunkIds/useGetOwnedPunkIds';

export { default as getOwnedWPunksIds } from './queries/getOwnedWPunksIds';
export * from './queries/getOwnedWPunksIds';
export { default as useGetOwnedWPunksIds } from './queries/getOwnedWPunksIds/useGetOwnedWPunksIds';

export { default as liquidateWithSingleRepay } from './mutations/liquidateWithSingleRepay';
export * from './mutations/liquidateWithSingleRepay';
export { default as useLiquidateWithSingleRepay } from './mutations/liquidateWithSingleRepay/useLiquidateWithSingleRepay';

export { default as liquidateBorrow } from './mutations/liquidateBorrow';
export * from './mutations/liquidateBorrow';
export { default as useLiquidateBorrow } from './mutations/liquidateBorrow/useLiquidateBorrow';

export { default as stakeXcn } from './mutations/stakeXcn';
export * from './mutations/stakeXcn';
export { default as useStakeXcn } from './mutations/stakeXcn/useStakeXcn';

export { default as withdrawXcn } from './mutations/withdrawXcn';
export * from './mutations/withdrawXcn';
export { default as useWithdrawXcn } from './mutations/withdrawXcn/useWithdrawXcn';

export { default as claimXcn } from './mutations/claimXcn';
export * from './mutations/claimXcn';
export { default as useClaimXcn } from './mutations/claimXcn/useClaimXcn';

export { default as getUniSwapPairs } from './queries/getUniSwapPairs';
export * from './queries/getUniSwapPairs';
export { default as useGetUniSwapPairs } from './queries/getUniSwapPairs/useGetUniSwapPairs';
