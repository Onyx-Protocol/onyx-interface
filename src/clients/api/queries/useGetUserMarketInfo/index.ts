import BigNumber from 'bignumber.js';
import { useMemo } from 'react';
import { Asset } from 'types';
import {
  calculateCollateralValue,
  convertTokensToWei,
  indexBy,
  unsafelyGetOToken,
  unsafelyGetToken,
} from 'utilities';

import {
  IGetOTokenBalancesAllOutput,
  useGetAssetsInAccount,
  useGetMarkets,
  useGetOTokenBalancesAll,
} from 'clients/api';
import { OETH_TOKENS, TOKENS } from 'constants/tokens';

export interface Data {
  assets: Asset[];
  userTotalBorrowLimitCents: BigNumber;
  userTotalBorrowBalanceCents: BigNumber;
  userTotalSupplyBalanceCents: BigNumber;
  totalXcnDistributedWei: BigNumber;
  dailyXcnWei: BigNumber;
}

export interface UseGetUserMarketInfoOutput {
  isLoading: boolean;
  data: Data;
}

const oTokenAddresses = Object.values(OETH_TOKENS).reduce(
  (acc, item) => (item.address ? [...acc, item.address] : acc),
  [] as string[],
);

// TODO: decouple, this hook handles too many things (see https://app.clickup.com/t/2d4rfx6)
const useGetUserMarketInfo = ({
  accountAddress,
}: {
  accountAddress?: string;
}): UseGetUserMarketInfoOutput => {
  const {
    data: getMarketsData = {
      markets: [],
      dailyXcnWei: new BigNumber(0),
    },
    isLoading: isGetMarketsLoading,
  } = useGetMarkets({
    placeholderData: {
      markets: [],
      dailyXcnWei: new BigNumber(0),
    },
  });

  const {
    data: assetsInAccount = {
      tokenAddresses: [],
    },
    isLoading: isGetAssetsInAccountLoading,
  } = useGetAssetsInAccount(
    { accountAddress: accountAddress || '' },
    {
      enabled: !!accountAddress,
      placeholderData: {
        tokenAddresses: [],
      },
    },
  );

  const {
    data: oTokenBalancesAccount = { balances: [] },
    isLoading: isGetOTokenBalancesAccountLoading,
  } = useGetOTokenBalancesAll(
    { account: accountAddress || '', oTokenAddresses },
    { enabled: !!accountAddress, placeholderData: { balances: [] } },
  );

  const oTokenBalances = useMemo(
    () =>
      indexBy(
        (item: IGetOTokenBalancesAllOutput['balances'][number]) => item.oToken.toLowerCase(), // index by oToken address
        oTokenBalancesAccount.balances,
      ),
    [JSON.stringify(oTokenBalancesAccount)],
  );

  const isLoading =
    isGetMarketsLoading || isGetAssetsInAccountLoading || isGetOTokenBalancesAccountLoading;

  const data = useMemo(() => {
    const {
      assets,
      userTotalBorrowBalanceCents,
      userTotalBorrowLimitCents,
      userTotalSupplyBalanceCents,
      totalXcnDistributedWei,
    } = (getMarketsData?.markets || []).reduce(
      (acc, market) => {
        const token = unsafelyGetToken(market.id);
        const oEthToken = unsafelyGetOToken(token.id);

        // Skip token if it isn't listed
        if (!token || !oEthToken) {
          return acc;
        }

        const otokenAddress = oEthToken.address.toLowerCase();
        const collateral = (assetsInAccount.tokenAddresses || [])
          .map((address: string) => address.toLowerCase())
          .includes(otokenAddress);

        let walletBalance = new BigNumber(0);
        let supplyBalance = new BigNumber(0);
        let borrowBalance = new BigNumber(0);
        const percentOfLimit = '0';

        const wallet = oTokenBalances && oTokenBalances[otokenAddress];
        if (accountAddress && wallet) {
          const toDecimalAmount = (mantissa: string) =>
            new BigNumber(mantissa).shiftedBy(-token.decimals);

          walletBalance = toDecimalAmount(wallet.tokenBalance);
          supplyBalance = toDecimalAmount(wallet.balanceOfUnderlying);
          borrowBalance = toDecimalAmount(wallet.borrowBalanceCurrent);
        }

        const asset: Asset = {
          token,
          supplyApy: new BigNumber(market?.supplyApy || 0),
          borrowApy: new BigNumber(market?.borrowApy || 0),
          xcnSupplyApy: new BigNumber(market?.supplyXcnApy || 0),
          xcnBorrowApy: new BigNumber(market?.borrowXcnApy || 0),
          collateralFactor: new BigNumber(market?.collateralFactor || 0).div(1e18),
          tokenPrice: new BigNumber(market?.tokenPrice || 0),
          liquidity: new BigNumber(market?.liquidity || 0),
          borrowCaps: new BigNumber(market?.borrowCaps || 0),
          supplyCaps: new BigNumber(market?.supplyCap || 0),
          treasuryTotalBorrowsCents: new BigNumber(market?.totalBorrowsUsd || 0).times(100),
          treasuryTotalSupplyCents: new BigNumber(market?.totalSupplyUsd || 0).times(100),
          treasuryTotalSupply: new BigNumber(market?.totalSupply || 0),
          treasuryTotalBorrows: new BigNumber(market?.totalBorrows2 || 0),
          cash: new BigNumber(market?.cash || 0),
          totalReserves: new BigNumber(market?.totalReserves || 0),
          walletBalance,
          supplyBalance,
          borrowBalance,
          collateral,
          percentOfLimit,
          xcnPerDay: new BigNumber(market?.supplierDailyXcn || 0)
            .plus(new BigNumber(market?.borrowerDailyXcn || 0))
            .div(new BigNumber(10).pow(TOKENS.xcn.decimals)),
        };

        // user totals
        const borrowBalanceCents = asset.borrowBalance.times(asset.tokenPrice).times(100);
        const supplyBalanceCents = asset.supplyBalance.times(asset.tokenPrice).times(100);
        acc.userTotalBorrowBalanceCents = acc.userTotalBorrowBalanceCents.plus(borrowBalanceCents);
        acc.userTotalSupplyBalanceCents = acc.userTotalSupplyBalanceCents.plus(supplyBalanceCents);

        acc.totalXcnDistributedWei = acc.totalXcnDistributedWei.plus(
          new BigNumber(market?.totalDistributed || 0).times(
            new BigNumber(10).pow(TOKENS.xcn.decimals),
          ),
        );

        // Create borrow limit based on assets supplied as collateral
        if (asset.collateral) {
          acc.userTotalBorrowLimitCents = acc.userTotalBorrowLimitCents.plus(
            calculateCollateralValue({
              amountWei: convertTokensToWei({ value: asset.supplyBalance, token }),
              token: asset.token,
              tokenPriceTokens: asset.tokenPrice,
              collateralFactor: asset.collateralFactor,
            }).times(100),
          );
        }

        return { ...acc, assets: [...acc.assets, asset] };
      },
      {
        assets: [] as Asset[],
        userTotalBorrowBalanceCents: new BigNumber(0),
        userTotalBorrowLimitCents: new BigNumber(0),
        userTotalSupplyBalanceCents: new BigNumber(0),
        totalXcnDistributedWei: new BigNumber(0),
      },
    );

    let assetList = assets;

    const userTotalBorrowBalance = userTotalBorrowBalanceCents;

    // percent of limit
    assetList = assetList.map((item: Asset) => ({
      ...item,
      percentOfLimit: new BigNumber(userTotalBorrowLimitCents).isZero()
        ? '0'
        : item.borrowBalance
            .times(item.tokenPrice)
            .div(userTotalBorrowLimitCents)
            .times(100)
            .dp(0, 1)
            .toFixed(),
    }));

    return {
      assets: assetList,
      userTotalBorrowBalanceCents: userTotalBorrowBalance,
      userTotalBorrowLimitCents,
      userTotalSupplyBalanceCents,
      dailyXcnWei: getMarketsData.dailyXcnWei || new BigNumber(0),
      totalXcnDistributedWei,
    };
  }, [
    '0',
    JSON.stringify(getMarketsData?.markets),
    JSON.stringify(assetsInAccount),
    JSON.stringify(oTokenBalances),
    JSON.stringify(getMarketsData),
  ]);

  return {
    isLoading,
    data,
    // TODO: handle errors and retry scenarios
  };
};

export default useGetUserMarketInfo;
