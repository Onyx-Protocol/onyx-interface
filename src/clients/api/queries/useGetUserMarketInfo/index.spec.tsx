import { waitFor } from '@testing-library/react';
import BigNumber from 'bignumber.js';
import React from 'react';

import fakeAddress from '__mocks__/models/address';
import { assetsInAccount } from '__mocks__/models/assetsInAccount';
import { markets } from '__mocks__/models/markets';
import { oTokenBalanceTreasury } from '__mocks__/models/oTokenBalanceTreasury';
import { oTokenBalancesAccount } from '__mocks__/models/oTokenBalancesAccount';
import { getAssetsInAccount, getMarkets, useGetOTokenBalancesAll } from 'clients/api';
import renderComponent from 'testUtils/renderComponent';

import useGetUserMarketInfo, { UseGetUserMarketInfoOutput } from '.';

jest.mock('clients/api');

describe('api/queries/useGetUserMarketInfo', () => {
  beforeEach(() => {
    (getMarkets as jest.Mock).mockImplementation(() => ({ markets }));
    (getAssetsInAccount as jest.Mock).mockImplementation(() => ({
      tokenAddresses: assetsInAccount,
    }));

    (useGetOTokenBalancesAll as jest.Mock).mockImplementation(({ account }) => {
      if (account === fakeAddress) {
        return { data: { balances: oTokenBalancesAccount } };
      }
      return { data: { balances: oTokenBalanceTreasury } };
    });
  });

  it('calculates totals correctly', async () => {
    let data: UseGetUserMarketInfoOutput['data'] = {
      assets: [],
      userTotalBorrowBalanceCents: new BigNumber(0),
      userTotalBorrowLimitCents: new BigNumber(0),
      userTotalSupplyBalanceCents: new BigNumber(0),
      totalXcnDistributedWei: new BigNumber(0),
      dailyXcnWei: new BigNumber(0),
    };

    const CallMarketContext = () => {
      ({ data } = useGetUserMarketInfo({ accountAddress: fakeAddress }));
      return <div />;
    };

    renderComponent(<CallMarketContext />, {
      authContextValue: { account: { address: fakeAddress } },
    });

    await waitFor(() => expect(data.assets.length > 0).toBe(true));
    expect(data).toMatchSnapshot();
  });
});
