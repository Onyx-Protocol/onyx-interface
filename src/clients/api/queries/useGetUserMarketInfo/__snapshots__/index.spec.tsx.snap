// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api/queries/useGetUserMarketInfo calculates totals correctly 1`] = `
Object {
  "assets": Array [
    Object {
      "borrowApy": "-2.0144969997738267",
      "borrowBalance": "0",
      "borrowCaps": "0",
      "cash": "1.0001000096016525033105349e+25",
      "collateral": false,
      "collateralFactor": "0.81",
      "liquidity": "10001830379044496325.01359740607398",
      "percentOfLimit": "0",
      "supplyApy": "0.000001539982091825",
      "supplyBalance": "0",
      "supplyCaps": "0",
      "token": Object {
        "address": "0x0397A41430153bFF2d7514fe1586f2e18E735869",
        "asset": "usdc.svg",
        "decimals": 6,
        "id": "usdc",
        "symbol": "USDC",
      },
      "tokenPrice": "1.00008302",
      "totalReserves": "47245578337903163",
      "treasuryTotalBorrows": "8586112918853.550179",
      "treasuryTotalBorrowsCents": "858682573794807.340073586058",
      "treasuryTotalSupply": "4.717199913187927152720001083e+28",
      "treasuryTotalSupplyCents": "1.0001838918443319696540393346289641043e+21",
      "walletBalance": "0",
      "xcnBorrowApy": "0.000002125348834693",
      "xcnPerDay": "999.9999936",
      "xcnSupplyApy": "1.824635e-12",
    },
    Object {
      "borrowApy": "-2.014488268288307",
      "borrowBalance": "0",
      "borrowCaps": "0",
      "cash": "1.009900000007157528256867649649e+30",
      "collateral": false,
      "collateralFactor": "0.8",
      "liquidity": "1.0097990100071568125040419628840351e+24",
      "percentOfLimit": "0",
      "supplyApy": "0",
      "supplyBalance": "0",
      "supplyCaps": "0",
      "token": Object {
        "address": "0xcE546eE9E8Bc0732A1e0083ef578a19F79c73D74",
        "asset": "usdt.svg",
        "decimals": 6,
        "id": "usdt",
        "symbol": "USDT",
      },
      "tokenPrice": "0.9999",
      "totalReserves": "5143616150",
      "treasuryTotalBorrows": "30465941.078326",
      "treasuryTotalBorrowsCents": "3046289448.42181674",
      "treasuryTotalSupply": "5.029972090817266864401527367893625e+33",
      "treasuryTotalSupplyCents": "1.009799010007156842907638340455919692236935e+26",
      "walletBalance": "0",
      "xcnBorrowApy": "0.600882672824011072",
      "xcnPerDay": "999.9999936",
      "xcnSupplyApy": "0",
    },
    Object {
      "borrowApy": "-2.8269802937000055",
      "borrowBalance": "0",
      "borrowCaps": "0",
      "cash": "9.706099325362378744807744335e+27",
      "collateral": false,
      "collateralFactor": "0.8",
      "liquidity": "9708012785.783380684153495733",
      "percentOfLimit": "0",
      "supplyApy": "0.222026939368167748",
      "supplyBalance": "0",
      "supplyCaps": "0",
      "token": Object {
        "address": "0xE39Ee8E20295572F7ED2Ea6Ae429f91697eCAaaF",
        "asset": "dai.svg",
        "decimals": 18,
        "id": "dai",
        "symbol": "DAI",
      },
      "tokenPrice": "1.00019714",
      "totalReserves": "0",
      "treasuryTotalBorrows": "838879295.456576767818085792",
      "treasuryTotalBorrowsCents": "83904467212.0883077362093449",
      "treasuryTotalSupply": "51881576366895517915",
      "treasuryTotalSupplyCents": "1054705745790.4263761515589172",
      "walletBalance": "0",
      "xcnBorrowApy": "0.021753286763848276",
      "xcnPerDay": "999.9999936",
      "xcnSupplyApy": "0.001730355354365121",
    },
    Object {
      "borrowApy": "-94.54098055818199",
      "borrowBalance": "0",
      "borrowCaps": "0",
      "cash": "134724956392949930",
      "collateral": false,
      "collateralFactor": "0.8",
      "liquidity": "39.742992883765334217",
      "percentOfLimit": "0",
      "supplyApy": "86.319696955441989068",
      "supplyBalance": "0",
      "supplyCaps": "0",
      "token": Object {
        "address": "",
        "asset": "eth.svg",
        "decimals": 18,
        "id": "eth",
        "isNative": true,
        "symbol": "ETH",
      },
      "tokenPrice": "294.99354795",
      "totalReserves": "6.133177004077521280386e+21",
      "treasuryTotalBorrows": "163570.361869451489735025",
      "treasuryTotalBorrowsCents": "4644293442.7935753968443612",
      "treasuryTotalSupply": "475703606186365",
      "treasuryTotalSupplyCents": "4644297417.0928637733777829",
      "walletBalance": "0",
      "xcnBorrowApy": "1.141107132664179593",
      "xcnPerDay": "2999.9999808",
      "xcnSupplyApy": "1.185821799653397884",
    },
    Object {
      "borrowApy": "-19.90282685098207",
      "borrowBalance": "0",
      "borrowCaps": "270000",
      "cash": "1.384009148852251251535718e+24",
      "collateral": false,
      "collateralFactor": "0.6",
      "liquidity": "1384009.148852251251535718",
      "percentOfLimit": "0",
      "supplyApy": "2.509901167951039272",
      "supplyBalance": "0",
      "supplyCaps": "0",
      "token": Object {
        "address": "0xc03Fa5e041BE9C8deD030D6dec7936cE67D78ae8",
        "asset": "xcn.svg",
        "decimals": 18,
        "id": "xcn",
        "symbol": "XCN",
      },
      "tokenPrice": "1",
      "totalReserves": "9.456129458228979467263e+21",
      "treasuryTotalBorrows": "306488.837102836344256144",
      "treasuryTotalBorrowsCents": "30648883.7102836344256144",
      "treasuryTotalSupply": "8238070729630362",
      "treasuryTotalSupplyCents": "168090555.8345572668784761",
      "walletBalance": "0",
      "xcnBorrowApy": "0",
      "xcnPerDay": "0",
      "xcnSupplyApy": "0",
    },
    Object {
      "borrowApy": "-5.7292945208867785",
      "borrowBalance": "0",
      "borrowCaps": "0",
      "cash": "734855505167845506968",
      "collateral": false,
      "collateralFactor": "0.8",
      "liquidity": "22080059.440703536158904017",
      "percentOfLimit": "0",
      "supplyApy": "1.814867084183460806",
      "supplyBalance": "0",
      "supplyCaps": "0",
      "token": Object {
        "address": "0x81969959E0485a38585C0e4cD3C8e4d996F12758",
        "asset": "wbtc.svg",
        "decimals": 8,
        "id": "wbtc",
        "symbol": "WBTC",
      },
      "tokenPrice": "30046.80414779",
      "totalReserves": "1766090250815625642",
      "treasuryTotalBorrows": "410.044985006090348771",
      "treasuryTotalBorrowsCents": "1232054135.626148384986397",
      "treasuryTotalSupply": "5387756610941",
      "treasuryTotalSupplyCents": "3434747710.1218290503961459",
      "walletBalance": "0",
      "xcnBorrowApy": "4.543731091576054636",
      "xcnPerDay": "2999.9999808",
      "xcnSupplyApy": "1.606740170918635611",
    },
    Object {
      "borrowApy": "-8.426578436066722",
      "borrowBalance": "0",
      "borrowCaps": "0",
      "cash": "4.442714417049498808402e+21",
      "collateral": false,
      "collateralFactor": "0.8",
      "liquidity": "9845948.33907734952053666",
      "percentOfLimit": "0",
      "supplyApy": "4.552151837121089047",
      "supplyBalance": "0",
      "supplyCaps": "0",
      "token": Object {
        "address": "",
        "asset": "eth.svg",
        "decimals": 18,
        "id": "eth",
        "isNative": true,
        "symbol": "ETH",
      },
      "tokenPrice": "2216.20104621",
      "totalReserves": "27361500255976792194",
      "treasuryTotalBorrows": "6946.982350412691554991",
      "treasuryTotalBorrowsCents": "1539590955.2987011849433085",
      "treasuryTotalSupply": "54560058223524",
      "treasuryTotalSupplyCents": "2518082368.1866852001814986",
      "walletBalance": "0",
      "xcnBorrowApy": "0.594444965549789125",
      "xcnPerDay": "500.0000256",
      "xcnSupplyApy": "0.363034531797658077",
    },
    Object {
      "borrowApy": "-3.3210326984806846",
      "borrowBalance": "0",
      "borrowCaps": "0",
      "cash": "1.366473497104608422125521e+24",
      "collateral": false,
      "collateralFactor": "0.6",
      "liquidity": "1096504.17926256565752608",
      "percentOfLimit": "0",
      "supplyApy": "0.250490295580310715",
      "supplyBalance": "0",
      "supplyCaps": "0",
      "token": Object {
        "address": "0x8207E64018f3043626E222D878beDCcb75348178",
        "asset": "matic.svg",
        "decimals": 18,
        "id": "matic",
        "symbol": "MATIC",
      },
      "tokenPrice": "0.80243355",
      "totalReserves": "367622752501696670187",
      "treasuryTotalBorrows": "127045.101864562779374701",
      "treasuryTotalBorrowsCents": "10194525.2099292730251508",
      "treasuryTotalSupply": "7442645752253233",
      "treasuryTotalSupplyCents": "119815375.5117972192153777",
      "walletBalance": "0",
      "xcnBorrowApy": "0",
      "xcnPerDay": "0",
      "xcnSupplyApy": "0",
    },
  ],
  "dailyXcnWei": "0",
  "totalXcnDistributedWei": "2.648826e+24",
  "userTotalBorrowBalanceCents": "0",
  "userTotalBorrowLimitCents": "0",
  "userTotalSupplyBalanceCents": "0",
}
`;
